dht/posix: rename fop handle a case of a stale linkto's gfid

tshacked opened the pull request:
Problem:
subvol-0: src located on the hashed subvol
subvol-1: stale linkto to src located on a non-hashed subvol (with a diff gfid)

During rename(src,dst): if a stale linkto to src is present in the dst
hashed, linkto creation (a stage in rename fop) fails with EEXIST.

Fix: 
populate the MKNOD_REPLACE key for "forcing" posix the creation of the linkto with correct gfid.


Change-Id: I07f4df1b7028ce15eb65bb5d60df1d5375c5243e
Updates: #1723
Signed-off-by: Tamar Shacked <tshacked@redhat.com>

tshacked committed:
dht/posix: rename fop handle case of stale linkto file (option B)

problem:
subvol-0: file1 located on the hashed subvol
subvol-1: stale linkto to file1 located on a non-hashed subvol (with a diff gfid)

During rename fop: file1-->file3, when the hashed location of file3 is subvol-1.
- dht attempt to create a linkto named file1 on subvol-1 failed with EEXIST.
- dht send lookup(file3) and due to gfid-changed afr unwind the fop with ESTALE

Fix:
Add internal flag to the linkto creation request.
At posix if the file present treat it as a stale link to and do unlink.

Change-Id: I07f4df1b7028ce15eb65bb5d60df1d5375c5243e
Updates: #1723
Signed-off-by: Tamar Shacked <tshacked@redhat.com>

xhernandez commented on xlators/cluster/dht/src/dht-common.h:
```diff
@@ -762,6 +762,23 @@ typedef struct dht_fd_ctx {
         }                                                                      \
     } while (0)
 
+#define DHT_MARK_RENAME_FOP(xattr)                                             \
```
Why a macro instead of a function ?

xhernandez replied:
This should be a function, not a macro. Using a function you can return an error in case xattr allocation or dict_set_str fails.

xhernandez commented on xlators/storage/posix/src/posix-entry-ops.c:
```diff
@@ -568,9 +568,15 @@ posix_mknod(call_frame_t *frame, xlator_t *this, loc_t *loc, mode_t mode,
         } else {
-            if (op_errno == EEXIST)
+            if (op_errno == EEXIST) {
+                if (dict_get_sizen(xdata, DHT_RENAME_FOP_KEY)) {
```
Instead of creating a new specific key, I would reuse the already existing `GLUSTERFS_INTERNAL_FOP_KEY` (checked at line 524). If it's present and we are here, it means that the requested gfid doesn't exist but the file exists. If I'm not wrong, the only callers that set `GLUSTERFS_INTERNAL_FOP_KEY` for an mknod operation are DHT, AFR, EC and Shard. In all cases it's to repair files (so the operation should succeed and it's not expected to find any existing file) or manipulate internal files. So I think it's safe to always delete the existing file in this case, without using a new specific key.

xhernandez commented on xlators/storage/posix/src/posix-entry-ops.c:
```diff
@@ -568,9 +568,15 @@ posix_mknod(call_frame_t *frame, xlator_t *this, loc_t *loc, mode_t mode,
+                if (dict_get_sizen(xdata, DHT_RENAME_FOP_KEY)) {
+                    dict_del_sizen(xdata, DHT_RENAME_FOP_KEY);
+                    op_ret = posix_unlink_stale_linkto(this, real_path);
+                    if (op_ret == 0)
```
If this fails, I would set log level to `GF_LOG_ERROR`.

xhernandez replied:
This is still relevant.

xhernandez commented on xlators/storage/posix/src/posix-handle.c:
```diff
@@ -979,3 +979,28 @@ posix_create_link_if_gfid_exists(xlator_t *this, uuid_t gfid, char *real_path,
+                     "mknod %s failed due stale linkto file, unlink",
+                     real_path);
+        /*neex to stale file xattr bla*/
+        ret = sys_unlink(real_path);
```
We also need to remove the related gfid file, but it could have more hardlinks. We should reuse part of the code from `posix_unlink()` to delete the existing file and, if necessary, it's gfid.

tshacked committed:
dht/posix: handle the deletion of the stale linkto's gfid

reuse posix_unlink_gfid_handle_and_entry() for deleting the stale linkto and it's gfid
wip: if this logic can be uses in all linkto creation paths including rebalance
     If we will use the new flag changed it to DHT_FORCE_MKNOD_KEY(?)

Updates: #1723

Change-Id: I33c5ba91860708a49563fc6c681fe0e66f7c85d4
Signed-off-by: Tamar Shacked <tshacked@redhat.com>

xhernandez commented on libglusterfs/src/glusterfs/glusterfs.h:
```diff
@@ -315,6 +315,7 @@ enum gf_internal_fop_indicator {
 #define DHT_SKIP_OPEN_FD_UNLINK "dont-unlink-for-open-fd"
 #define DHT_IATT_IN_XDATA_KEY "dht-get-iatt-in-xattr"
 #define DHT_MODE_IN_XDATA_KEY "dht-get-mode-in-xattr"
+#define DHT_RENAME_FOP_KEY "dht-rename-fop"
```
I would use a more generic name unrelated to DHT, like `GF_MKNOD_REPLACE_KEY` for example.

tshacked replied:
done

xhernandez commented on xlators/cluster/dht/src/dht-rename.c:
```diff
@@ -1385,12 +1392,20 @@ dht_rename_create_links(call_frame_t *frame)
-                     dst_hashed->name, src_cached->name);
+        dict_t *xattr_new = NULL;
+
+        xattr_new = dict_copy_with_ref(xattr, NULL);
```
Needs error checking.

xhernandez commented on xlators/cluster/dht/src/dht-rename.c:
```diff
@@ -1385,12 +1392,20 @@ dht_rename_create_links(call_frame_t *frame)
+
+        xattr_new = dict_copy_with_ref(xattr, NULL);
+        DHT_MARK_RENAME_FOP(xattr_new);
+        local->params = dict_ref(xattr_new);
```
Instead of taking a new ref we can simply assign `xattr_new` here and remove the unref at line 1408.

xhernandez commented on xlators/storage/posix/src/posix-entry-ops.c:
```diff
@@ -1184,6 +1196,42 @@ posix_unlink_gfid_handle_and_entry(call_frame_t *frame, xlator_t *this,
+               "lstat on %s failed", real_path);
+        goto out;
+    }
+    /* unlink only if linkto filw*/
```
```suggestion
    /* unlink only if linkto file*/
```

xhernandez commented on xlators/storage/posix/src/posix-entry-ops.c:
```diff
@@ -1184,6 +1196,42 @@ posix_unlink_gfid_handle_and_entry(call_frame_t *frame, xlator_t *this,
+            frame, this, real_path, &stbuf, op_errno, loc, _gf_false, NULL);
+
+    } else {
+        gf_msg(this->name, GF_LOG_WARNING, 0, P_MSG_HANDLE_CREATE,
```
Ok for debugging, but I think it should be removed in the final patch (or keep it in debug level).

xhernandez commented on xlators/storage/posix/src/posix-entry-ops.c:
```diff
@@ -1184,6 +1196,42 @@ posix_unlink_gfid_handle_and_entry(call_frame_t *frame, xlator_t *this,
+    /* get stale file gfid and stat-info */
+    ret = posix_pstat(this, NULL, NULL, real_path, &stbuf, _gf_false);
+    if (ret) {
+        if (errno == ENOENT)
```
We can only be here if the file existed. This shouldn't happen. It means we have a race somewhere. We should add a log message here.

tshacked replied:
Added INFO logging, 
One legit race can be with rebalance.

tshacked committed:
dht/posix rename-path: handle the deletion of the stale linkto's gfid

during rename(src,dst): if a stale linkto to src is present in the dst
hashed, linkto creation fails with EEXIST.

handling: populate the MKNOD REPLACE key for "forcing" posix the
creation of the linkto with correct gfid.

Change-Id: I02d2023b213a19ae7e9a934e3375a3bddaa179c9
Signed-off-by: Tamar Shacked <tshacked@redhat.com>

tshacked commented:
/run regression

xhernandez commented on libglusterfs/src/glusterfs/glusterfs.h:
```diff
@@ -315,6 +315,7 @@ enum gf_internal_fop_indicator {
 #define DHT_SKIP_OPEN_FD_UNLINK "dont-unlink-for-open-fd"
 #define DHT_IATT_IN_XDATA_KEY "dht-get-iatt-in-xattr"
 #define DHT_MODE_IN_XDATA_KEY "dht-get-mode-in-xattr"
+#define GF_MKNOD_REPLACE_KEY "mknod-replace-if-linkto"
```
Why not just this:

```suggestion
#define GF_FORCE_REPLACE_KEY "force-replace"
```

This way the name is more generic and it could be easily reused in the future if necessary for other purposes not related to linkto files, and for other fops, not only mknod.

xhernandez commented on xlators/cluster/dht/src/dht-rename.c:
```diff
@@ -1385,12 +1386,36 @@ dht_rename_create_links(call_frame_t *frame)
+                   "Failed to set dictionary value: key = %s,"
+                   " path = %s",
+                   GF_MKNOD_REPLACE_KEY, local->loc.path);
+            ret = 0;
```
Are you sure we can ignore this error ? in most of the cases the linkto file won't exist, so the presence of this key is irrelevant. But when the linkto file exists, failing to set this key will cause failure of the rename. I think it's better to fail here with a more consistent error, like ENOMEM, than a later failure with ESTALE or EEXIST which is harder to understand by the user.

We also fail if `xattr_new` cannot be allocated, so I think we should also fail if the key cannot be added.

tshacked replied:
Actually, I had doubts about it, as it handles rare cases i was not sure if we should fail the fop.
But i agree that ENOMEM is a more consistent error and a failure in the infrastructure should be rare too..

xhernandez commented on xlators/storage/posix/src/posix-entry-ops.c:
```diff
@@ -1184,6 +1200,47 @@ posix_unlink_gfid_handle_and_entry(call_frame_t *frame, xlator_t *this,
+    } else {
+        gf_msg(this->name, GF_LOG_DEBUG, 0, P_MSG_HANDLE_CREATE,
+               "skip unlinking stale data-file: %s gfid: %s", real_path,
+               uuid_utoa(stbuf.ia_gfid));
```
We need to return some error in this case (`ret` is 0 here).

tshacked replied:
looks fine to me, it's being initialized: `int ret = -1;`

xhernandez replied:
Initialization value is overwritten at line 1213 and we only get here if `ret` was 0.

tshacked committed:
dht/posix rename-path: handle the deletion of the stale linkto's gfid

during rename(src,dst): if a stale linkto to src is present in the dst
hashed, linkto creation fails with EEXIST.

handling: populate the MKNOD REPLACE key for "forcing" posix the
creation of the linkto with correct gfid.

Updates: #1723

Change-Id: Idd08920090a8ef52e792bb94f86cd9f47d10c15a
Signed-off-by: Tamar Shacked <tshacked@redhat.com>

tshacked committed:
Fixing code review comments

Change-Id: I4a62049babfd11f14ffff37746134b9089655032
Signed-off-by: Tamar Shacked <tshacked@redhat.com>

tshacked commented:
/run regression

mohit84 reviewed (commented):
LGTM +1

xhernandez commented on libglusterfs/src/glusterfs/glusterfs.h:
```diff
@@ -315,7 +315,7 @@ enum gf_internal_fop_indicator {
 #define DHT_IATT_IN_XDATA_KEY "dht-get-iatt-in-xattr"
 #define DHT_MODE_IN_XDATA_KEY "dht-get-mode-in-xattr"
-#define GF_MKNOD_REPLACE_KEY "mknod-replace-if-linkto"
+#define GF_MKNOD_REPLACE_KEY "force-replace"
```
The name of the macro should also be changed.

tshacked committed:
dht/posix rename-path: handle the deletion of the stale linkto's gfid

handle code review comments

Updates: #1723

Change-Id: Ica24b9a3ac3b7b50607897bbec10aa7ab3f03a0d
Signed-off-by: Tamar Shacked <tshacked@redhat.com>

tshacked commented:
/run regression
